packages/ssz/src/type/optional.ts

```diff
@@ -0,0 +1,224 @@
+    return value !== null ? 1 + this.elementType.value_serializedSize(value) : 0;
+  }
+
+  value_serializeToBytes(output: ByteViews, offset: number, value: ValueOfType<ElementType>): number {
+    if (value !== null) {
+      output.uint8Array[offset] = 1;
+      return this.elementType.value_serializeToBytes(output, offset + 1, value);
+    } else {
+      return offset;
+    }
+  }
+
+  value_deserializeFromBytes(data: ByteViews, start: number, end: number): ValueOfType<ElementType> {
+    if (start === end) {
+      return null as ValueOfType<ElementType>;
+    } else {
+      const selector = data.uint8Array[start];
+      if (selector !== 1) {
+        throw Error(`Invalid selector=${selector} for Optional type`);
+      }
+      return this.elementType.value_deserializeFromBytes(data, start + 1, end) as ValueOfType<ElementType>;
+    }
+  }
+
+  tree_serializedSize(node: Node): number {
+    const length = getLengthFromRootNode(node);
+    return length === 1 ? 1 + this.elementType.value_serializedSize(node.left) : 0;
+  }
+
+  tree_serializeToBytes(output: ByteViews, offset: number, node: Node): number {
+    const selector = getLengthFromRootNode(node);
+
+    const valueNode = node.left;
+    if (selector === 0) {
+      return offset;
+    } else {
+      output.uint8Array[offset] = 1;
+    }
+    return this.elementType.tree_serializeToBytes(output, offset + 1, valueNode);
+  }
+
+  tree_deserializeFromBytes(data: ByteViews, start: number, end: number): Node {
+    let valueNode;
+    let selector;
+    if (start === end) {
+      selector = 0;
+      valueNode = zeroNode(0);
+    } else {
+      selector = data.uint8Array[start];
+      if (selector !== 1) {
+        throw Error(`Invalid selector=${selector} for Optional type`);
+      }
+      valueNode = this.elementType.tree_deserializeFromBytes(data, start + 1, end);
+    }
+
+    return addLengthNode(valueNode, selector);
+  }
+
+  // Merkleization
+
+  hashTreeRoot(value: ValueOfType<ElementType>): Uint8Array {
+    const selector = value === null ? 0 : 1;
+    return mixInLength(super.hashTreeRoot(value), selector);
+  }
+
+  protected getRoots(value: ValueOfType<ElementType>): Uint8Array[] {
+    const valueRoot = value ? this.elementType.hashTreeRoot(value) : new Uint8Array(32);
+    return [valueRoot];
+  }
+
+  // Proofs
+
+  getPropertyGindex(prop: string): bigint {
+    if (isCompositeType(this.elementType)) {
+      const propIndex = this.elementType.getPropertyGindex(prop);
+      if (propIndex === null) {
+        throw Error(`index not found for property=${prop}`);
+      }
+      return concatGindices([VALUE_GINDEX, propIndex]);
+    } else {
+      throw new Error("not applicable for Optional basic type");
+    }
+  }
+
+  getPropertyType(): Type<unknown> {
+    return this.elementType;
+  }
+
+  getIndexProperty(index: number): string | number | null {
+    if (isCompositeType(this.elementType)) {
+      return this.elementType.getIndexProperty(index);
```

Comment:
may be an de-concatenation op is required (to remove `VALUE_GINDEX` before passing in the index to the underlying element type

Reply:
I think this method should only return `"value"` or `"selector"` for 0 or 1?

Reply:
transparent because `value` and `selector` properties don't exist as per my understanding